cmd/dryrun/dryrun.go

```diff
@@ -0,0 +1,535 @@
+
+	if err := k8syaml.UnmarshalStrict(policyBytes, &unstruct.Object); err != nil {
+		return nil, err
+	}
+
+	if unstruct.GetAPIVersion() != "policy.open-cluster-management.io/v1" {
+		return nil, fmt.Errorf("unsupported apiVersion: %v, the input policy must be "+
+			"'policy.open-cluster-management.io/v1'", unstruct.GetAPIVersion())
+	}
+
+	cfgpol := policyv1.ConfigurationPolicy{}
+
+	switch unstruct.GetKind() {
+	case "ConfigurationPolicy":
+		if err := k8syaml.UnmarshalStrict(policyBytes, &cfgpol); err != nil {
+			return nil, fmt.Errorf("could not unmarshal input to a ConfigurationPolicy: %w", err)
+		}
+	case "Policy":
+		tmpls, found, err := unstructured.NestedSlice(unstruct.Object, "spec", "policy-templates")
+		if err != nil {
+			return nil, fmt.Errorf("invalid input Policy: %w", err)
+		}
+
+		if !found {
+			return nil, fmt.Errorf("invalid input Policy: no policy-templates found")
+		}
+
+		cfgPolFound := false
+
+		for i, tmpl := range tmpls {
+			tmplMap, ok := tmpl.(map[string]any)
+			if !ok {
+				continue
+			}
+
+			objDef, ok := tmplMap["objectDefinition"]
+			if !ok {
+				continue
+			}
+
+			objDefMap, ok := objDef.(map[string]any)
+			if !ok {
+				continue
+			}
+
+			if objDefMap["kind"] != "ConfigurationPolicy" {
+				continue
+			}
+
+			if cfgPolFound {
+				cmd.Println("Ignoring additional ConfigurationPolicy in input policy")
+
+				continue
+			}
+
+			cfgPolFound = true
+
+			cfgpolBytes, err := json.Marshal(objDef)
+			if err != nil {
+				return nil, fmt.Errorf("unable to marshal policy template back to JSON")
+			}
+
+			if err := k8syaml.UnmarshalStrict(cfgpolBytes, &cfgpol); err != nil {
+				return nil, fmt.Errorf("could not unmarshal input policy template [%v] to a "+
+					"ConfigurationPolicy: %w", i, err)
+			}
+		}
+
+		if !cfgPolFound {
+			return nil, fmt.Errorf("invalid input Policy: it must contain a ConfigurationPolicy")
+		}
+	default:
+		return nil, fmt.Errorf("unsupported input kind: %v, must be 'Policy' or 'ConfigurationPolicy'",
+			unstruct.GetKind())
+	}
+
+	cfgpol.Spec.RemediationAction = policyv1.Inform
+	cfgpol.Spec.EvaluationInterval = policyv1.EvaluationInterval{
+		Compliant:    "10s",
+		NonCompliant: "10s",
+	}
+
+	cfgpol.OwnerReferences = []metav1.OwnerReference{{
+		Name: parentName,
+	}}
+
+	return &cfgpol, nil
+}
+
+// readInputResources takes stdin and any paths given as "positional" arguments,
+// and decodes them from YAML into k8s resources. Directories can be passed in
+// arguments: in this case all files in that directory will be read and decoded
+// (but no "deeper" directories are considered).
+func (d *DryRunner) readInputResources(cmd *cobra.Command, args []string) (
+	[]*unstructured.Unstructured, error,
+) {
+	stdinInfo, err := os.Stdin.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("could not read stdin: %w", err)
+	}
+
+	rawInputs := []*unstructured.Unstructured{}
+
+	filepaths := []string{}
+
+	if stdinInfo.Mode()&os.ModeCharDevice == 0 {
```

Comment:
Not entirely sure if this is the correct fix, but when running it through the debugger where it has no stdin, it would just hang forever for the YAML decoder to get something from stdin.
```suggestion
	if stdinInfo.Mode()&os.ModeCharDevice == 0 && stdinInfo.Size() != 0 {
```

Reply:
We'll need to look at this more. With `&& stdinInfo.Size() != 0`, I found that the cli doesn't successfully read inputs from stdin anymore when you run it directly. The tests still pass, I think because I kind of hacked the inputs for the tests...

So based on the `test_multiple_resources` case:
```
❯ cat input_stdin.yaml | ./dryrun-original -p policy.yaml input_3.yaml input_4.yaml 
# Diffs:
v1 Pod default/nginx-pod-e2e:

v1 Pod default/nginx-pod-e2e-2:

v1 Pod default/nginx-pod-e2e-3:

v1 Pod default/nginx-pod-e2e-4:

# Compliance messages:
Compliant; notification - pods [nginx-pod-e2e] found as specified in namespace default; notification - pods [nginx-pod-e2e-2] found as specified in namespace default; notification - pods [nginx-pod-e2e-3] found as specified in namespace default; notification - pods [nginx-pod-e2e-4] found as specified in namespace default

❯ cat input_stdin.yaml | ./dryrun-sizene0 -p policy.yaml input_3.yaml input_4.yaml                 
# Diffs:
v1 Pod default/nginx-pod-e2e-3:

v1 Pod default/nginx-pod-e2e-4:

# Compliance messages:
Compliant; notification - pods [nginx-pod-e2e-3] found as specified in namespace default; notification - pods [nginx-pod-e2e-4] found as specified in namespace default
```

Reply:
Is there a debugger setting where you can set stdin?

Reply:
@JustinKuli I don't know of one but this is the launch.json section that I had with this issue:

```
        {
            "name": "Launch dryrun",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "program": "${workspaceFolder}/cmd/main.go",
            "args": [
                "-p",
                "${workspaceFolder}/cmd/dryrun/testdata/test_basic_compliant/policy.yaml",
                "${workspaceFolder}/cmd/dryrun/testdata/test_basic_compliant/input_1.yaml"
            ]
        }
```

Reply:
We can leave that out for now though!